### PR TITLE
Decommit memory as needed

### LINKED ISSUE #79882: Array might not be zero-initialized
### Description

We have found, that elements of an array might not be always initialized to the default value. For example, after calling new double[size], some elements might be nonzero. We have started noticing this problem after upgrading to .net7. It seems, that this happens especially when the GC heap size limit is set. (Initially, we observed this with 8 GB limit.)

### Reproduction Steps

Run this C#  program:
```c#
using System;

var rand = new Random();
try
{
    for (int k = 0; k < 1_000_000; k++)
    {
        Test(k);
    }
    Console.Write("Succeeded");
}
catch (Exception ex)
{
    Console.WriteLine("Failed");
    Console.WriteLine(ex.Message);    
}

void Test(int iteration)
{
    int size = (int)Math.Pow(10, 5 + rand.NextDouble() * (1 + rand.NextDouble()));

    var m = new double[size];
    for (int i = 0; i < size; i++)
        if (m[i] != 0)
            throw new Exception(m[i].ToString() + "; " + i + "/" + size + "; iteration " + iteration.ToString());

    for (int i = 0; i < size; i++)
        m[i] = 21;
}
```

With GC HeapHardLimit set to 1 GB
```
{
  "configProperties": {
    "System.GC.HeapHardLimit": 1000000000
  }
}
```

### Expected behavior

We expect that Exception is not thrown, because elements of the array should be initialized to zero.

### Actual behavior

With high probability, Exception is thrown, because an element is not zero.

### Regression?

We haven't noticed this issue in .net6.

### Known Workarounds

_No response_

### Configuration

.net version 7.0.100
MS Windows 10 Pro, x64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by cshung**
This change makes sure:

1. The mark array is decommitted when we decommit the region, and
2. The region is decommitted if we fail to commit the corresponding mark array.

Fixes https://github.com/dotnet/runtime/issues/79882

### CODE REVIEW

File: src/coreclr/gc/gc.cpp [hunk 1]
```diff
@@ -20325,22 +20325,31 @@ bool gc_heap::try_get_new_free_region()
...
+        // We don't have memory to commit the mark array so we cannot use the new region.
+        int h_number =
+#ifdef MULTIPLE_HEAPS
+            heap_number;
+#else
+            0;
+#endif
+        gc_oh_num oh = gen_to_oh (gen_number);
+        size_t size = heap_segment_committed(region) - heap_segment_mem (region);
+        virtual_decommit(heap_segment_mem (region), size, oh, h_number);
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:20345]: you shouldn't be decommitting regions if large pages are enabled though in `init_table_for_region`. I suggest to refactor what `decommit_step` is doing and call that instead.

File: src/coreclr/gc/gc.cpp [hunk 2]
```diff
@@ -20325,22 +20325,31 @@ bool gc_heap::try_get_new_free_region()
...
-        if ((region->flags & heap_segment_flags_ma_committed) != 0)
-        {
-            bgc_verify_mark_array_cleared (region, true);
-        }
+        // We don't have memory to commit the mark array so we cannot use the new region.
+        int h_number =
+#ifdef MULTIPLE_HEAPS
+            heap_number;
+#else
+            0;
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:20341]: `heap_number` is already defined as 0 for WKS GC

File: src/coreclr/gc/gc.cpp (hunk 1 above)
Comment (Maoni0) [reply src/coreclr/gc/gc.cpp:20345]: nit - you have inconsistent space in these 3 lines of code. please try to stick to having a space to the left of '(`.

File: src/coreclr/gc/gc.cpp [hunk 3]
```diff
@@ -41072,6 +41081,12 @@ bool gc_heap::decommit_step (uint64_t step_milliseconds)
             if (!use_large_pages_p)
             {
                 decommit_succeeded_p = virtual_decommit(page_start, size, recorded_committed_free_bucket);
+#ifdef MULTIPLE_HEAPS
+                gc_heap* hp = heap_segment_heap (region);
+#else
+                gc_heap* hp = __this;
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:41087]: nit - conventional way is to do

`gc_heap* hp = pGenGCHeap;`

for WKS GC

File: src/coreclr/gc/gc.cpp [hunk 4]
```diff
@@ -41072,6 +41081,12 @@ bool gc_heap::decommit_step (uint64_t step_milliseconds)
...
+#endif
+                hp->decommit_mark_array_by_seg (region);
```
Comment (Maoni0) [inline src/coreclr/gc/gc.cpp:41089]: why are we not doing the mark array decommit for the large pages case? wouldn't you hit the same problem with double counting mark array committed bytes?

**Review by Maoni0 — CHANGES_REQUESTED** [review]
also it would be nice to add a comment to `delete_region` to specify the invariant which is when you return a region to the global region allocator, memory should be decommitted (unless it's large pages) and mark array (if committed).

**Comment by PeterSolMS** [comment]
Looks good to me.

Regarding Maoni's comment, I wonder if we can add an assert to `delete_region` to actually enforce the invariant. To make that work, `decommit_region` would need to set `heap_segment_committed` for the region, so we know its memory is decommitted.

**Review by Maoni0 — APPROVED** [review]
LGTM!

**Comment by sebastienros** [comment]
With these commits (https://github.com/dotnet/runtime/compare/2ca7cf7...6aaaaaa) aspnet apps crash at runtime in the benchmarks CI. I am flagging the PR which had the potential to explain it.

Crank command if you can run with and without this PR changes:

```
crank --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/src/BenchmarksApps/Mvc/benchmarks.certapi.yml  --config https://raw.githubusercontent.com/aspnet/Benchmarks/main/build/ci.profile.yml --scenario NoMvcAuth --profile intel-win-app --profile intel-load2-load --profile intel-db-db --application.framework net8.0 --application.aspNetCoreVersion 8.0.0-alpha.1.23062.23 --application.runtimeVersion 8.0.0-alpha.1.23061.12 --application.sdkVersion 8.0.100-alpha.1.23061.8
```

**Pull request merged**